transport_plugins/websocket/iotile_transport_websocket/device_adapter.py

```diff
@@ -102,6 +155,281 @@ def stop_sync(self):
         self.client.stop()
         self.connections.stop()
 
+    def send_rpc_async(self, connection_id, address, rpc_id, payload, timeout, callback):
+        """Asynchronously send an RPC to this IOTile device
+
+        Args:
+            connection_id (int): A unique identifier that will refer to this connection
+            address (int): the address of the tile that we wish to send the RPC to
+            rpc_id (int): the 16-bit id of the RPC we want to call
+            payload (bytes): the payload of the command
+            timeout (float): the number of seconds to wait for the RPC to execute
+            callback (callable): A callback for when we have finished the RPC.  The callback will be called as"
+                callback(connection_id, adapter_id, success, failure_reason, status, payload)
+                'connection_id': the connection id
+                'adapter_id': this adapter's id
+                'success': a bool indicating whether we received a response to our attempted RPC
+                'failure_reason': a string with the reason for the failure if success == False
+                'status': the one byte status code returned for the RPC if success == True else None
+                'payload': a bytearray with the payload returned by RPC if success == True else None
+        """
+
+        try:
+            self.connections.get_connection_id(connection_id)
+        except ArgumentError:
+            callback(connection_id, self.id, False, "Could not find connection information", None, None)
+            return
+
+        self.connections.begin_operation(connection_id, 'rpc', callback, timeout)
+
+        try:
+            result = self.client.send_command('send_rpc', {
```

Comment:
As discussed yesterday, this is not truly async so you would need to post a message and then have the rest of this routine run in a callback on receiving the response back (timeout is handled by the connection manager).